### PR TITLE
fix(datastore): add missing errorHandler errors for sync and mutation unauth responses 

### PR DESCRIPTION
**Pull request by svidgen**
#### Description of changes

Adds missing `errorHandler` events for "top level" unauthorized errors in AppSync resolvers (i.e., when resolver VTL calls `$util.unauthorized()`).

To support the testing for this, this PR adds:

1. Ability to hook into `errorHandler` using an `Observable` from the `getDatastore()` test factory.
2. Ability to inject a single `intercept` method into the graphql fake, so that individual responses can be faked in the context of a full sync. (Helpful if/when we clamp on what a single failed sync, mutation, or subscription means to the overall state of DataStore.)

#### Issue #, if available

#11042

#### Description of how you validated changes

Added protocol tests based on:

1. Forcing relevant resolvers to call `$util.unauthorize`.
3. Capturing responses.
4. Added simulated responses to protocol test.
5. Add assertion for `errorHandler` to trigger an `Unauthorized` error.
6. Made the tests pass.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/src/sync/processors/sync.ts [hunk 1]
```diff
@@ -389,6 +414,11 @@ class SyncProcessor {
 											onTerminate
 										));
 									} catch (error) {
+										console.log(
+											'received error',
+											error,
+											getSyncErrorType(error)
+										);
```
Comment (svidgen) [inline packages/datastore/src/sync/processors/sync.ts:421]: Whoops! This is troubleshooting cruft. Will remove.

File: packages/datastore/__tests__/connectivityHandling.test.ts [hunk 2]
```diff
@@ -26,10 +26,22 @@ async function waitForEmptyOutboxOrError(service) {
 	return await Promise.race([waitForEmptyOutbox(), pendingError]);
 }
 
+function waitForNextErrorHandlerError(
```
Comment (svidgen) [inline packages/datastore/__tests__/connectivityHandling.test.ts:29]: Will rev with a docstring.

Comment (svidgen) [reply packages/datastore/__tests__/connectivityHandling.test.ts:29]: Also generalizing the name. This doesn't need to be specific to the `errorHandler` wrapper.

File: packages/datastore/src/sync/processors/sync.ts [hunk 3]
```diff
@@ -284,20 +285,44 @@ class SyncProcessor {
...
+							message: error.message,
+							model: modelDefinition.name,
+							operation: opName,
+							errorType: getSyncErrorType(error.errors[0]),
+							process: ProcessName.sync,
+							remoteModel: null!,
+							cause: error,
+						});
+						throw new NonRetryableError(error);
 					}
```
Comment (svidgen) [inline packages/datastore/src/sync/processors/sync.ts:326]: I need focus here.

I'm not aware of a partial data scenario that comes with auth errors. My impression is that partial data with unauthorized records does not generate *errors*, but instead simply filters `items`. If I'm wrong, I'd like someone to how to produce that response.

**Review by dpilch — COMMENTED** [review]
Is `errorHandler` not needed on the subscription processor?

**Comment by svidgen** [comment]
> Is `errorHandler` not needed on the subscription processor?

Subscriptions look like they are already issuing error handler messages for the unauth case (observed in studio watching browser console). I can submit a separate PR to clamp on that at the same level.

I don't want to delay the fix for the tests on the not-broken part, especially since I haven't actually looked at how those errors look coming though the `graphql` call yet. I need to do some manual testing to ensure I'm clamping the right interaction there.

**Review by dpilch — DISMISSED** [review]

**Review by dpilch — APPROVED** [review]

**Review by iartemiev — APPROVED** [review]

**Pull request merged**